allocator: carve out a package boundary

irfansharif opened the pull request:
**allocator/storepool: carve out package from kvserver**

This commit tries to pave the way for the allocator itself to be carved
out (storepool probably being a direct dependency, but ideally through
an interface). This commit shamelessly exports any and everything needed
in order for StorePool to sit outside the kvserver package boundary.
It's unfortunate how tangled up this component is, including in tests
(some of which reach deep into inner mutexes and mutate state). We're
not doing anything to improve the status quo other than moving it and
maybe pointing to the awkwardness.

**allocatorimpl: carve out an allocator package**

Do the same for kvserver/replicastats, a type that's used as an input
for the allocator at various points (needed to break the dependency
between the allocatorimpl and kvserver). We take the same approach as
the last commit, exporting anything and everything needed to introduce a
hard package boundary.

irfansharif committed:
allocator/storepool: carve out package from kvserver

This commit tries to pave the way for the allocator itself to be carved
out (storepool probably being a direct dependency, but ideally through
an interface). This commit shamelessly exports any and everything needed
in order for StorePool to sit outside the kvserver package boundary.
It's unfortunate how tangled up this component is, including in tests
(some of which reach deep into inner mutexes and mutate state). We're
not doing anything to improve the status quo other than moving it and
maybe pointing to the awkwardness.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79991)
<!-- Reviewable:end -->

irfansharif committed:
allocatorimpl: carve out an allocator package

Do the same for kvserver/replicastats, a type that's used as an input
for the allocator at various points (needed to break the dependency
between the allocatorimpl and kvserver). We take the same approach as
the last commit, exporting anything and everything needed to introduce a
hard package boundary.

Release note: None

kvoli commented on pkg/kv/kvserver/allocator/base.go:
```diff
@@ -0,0 +1,48 @@
+	// is greater than this value, it will never be used as a rebalance or
+	// allocate target and we will actively try to move replicas off of it.
+	MaxFractionUsedThreshold = 0.95
+)
```
Why did these move into a separate file? There are other similar constraint checks that should probably also tag along if these do get split here.

nit: the const declaration should be above the func.

irfansharif replied:
Everything that's shared between the subpackages found its way below into this base package, everything that could be scoped to just the subpackage was kept within. MaxCapacityCheck is used by both the allocatorimpl and storepool, which is why only these constants feature here. These boundaries are going to be fuzzy/awkward for now, it's not clear what things should be shared vs. not.

Moved the const declarations above.

kvoli commented on pkg/kv/kvserver/allocator_scorer.go:
```diff
@@ -46,16 +48,11 @@ const (
-	// or allocate target and we will actively try to move replicas off of it.
-	maxFractionUsedThreshold = 0.95
-
 	// rebalanceToMaxFractionUsedThreshold: if the fraction used of a store
```
This should also go along to base.go, if you have the other disk fullness check there.

irfansharif replied:
See comment above.

kvoli commented on pkg/kv/kvserver/allocator_scorer.go:
```diff
@@ -2082,11 +2064,6 @@ func (o storeHealthOptions) rebalanceToReadAmpIsHealthy(
-	return store.Capacity.FractionUsed() < maxFractionUsedThreshold
-}
-
 // rebalanceToMaxCapacityCheck returns true if the store has enough room to
```
Should this also go along to base.go, with the other disk fullness check? I'm don't think there's as much benefit from splitting this off from the scorer yet.

irfansharif replied:
See comment above.

irfansharif commented:
🍺 

bors r+
